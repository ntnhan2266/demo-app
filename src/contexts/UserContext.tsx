import React, { createContext, useState, ReactNode } from 'react';
import { PublicUser } from '@/interfaces/user';

// Context interface
interface IUserContext {
  user: PublicUser | null;
  setUser: (_user: PublicUser | null) => void;
}

// Create the context
export const UserContext = createContext<IUserContext | undefined>(undefined);

// Provider component
export const UserProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<PublicUser | null>(null);

  const contextValue: IUserContext = {
    user,
    setUser,
  };

  return <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>;
};
