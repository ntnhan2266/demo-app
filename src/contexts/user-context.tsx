import React, { createContext, useContext, useState, ReactNode } from 'react';
import { PublicUser } from '@/interfaces/user';

// Context interface
interface IUserContext {
  user: PublicUser | null;
  setUser: (user: PublicUser | null) => void;
}

// Create the context
const UserContext = createContext<IUserContext | undefined>(undefined);

// Provider component
export const UserProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<PublicUser | null>(null);

  const contextValue: IUserContext = {
    user,
    setUser,
  };

  return <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>;
};

// Custom hook to use the user context
export const useUser = () => {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};
