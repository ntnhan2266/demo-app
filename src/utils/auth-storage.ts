import { LOCALSTORAGE_KEY } from '@/constants/localstorage-key';
import { IUser } from '@/interfaces/user';

// Convert user data to base64 string format
export const saveAuthUserToLocalStorage = (authUser: IUser): void => {
  const userString = btoa(JSON.stringify(authUser));
  localStorage.setItem(LOCALSTORAGE_KEY.AUTH_USER, userString);
};

export const getAuthUserFromLocalStorage = (): IUser | null => {
  const userString = localStorage.getItem(LOCALSTORAGE_KEY.AUTH_USER);

  if (userString) {
    try {
      // Decode the Base64-encoded string and parse it as JSON
      const decodedUserString = atob(userString);
      const authUser = JSON.parse(decodedUserString) as IUser;

      return authUser;
    } catch (error) {
      console.error('Error decoding or parsing user data:', error);
    }
  }

  return null;
};

export const removeAuthUserFromLocalStorage = (): void => {
  localStorage.removeItem(LOCALSTORAGE_KEY.AUTH_USER);
};
