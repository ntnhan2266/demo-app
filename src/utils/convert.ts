type TransformFunction = (_key: string) => string;

const transformKeys = <T>(obj: Record<string, T>, keyTransform: TransformFunction): Record<string, T> => {
  if (typeof obj !== 'object' || obj === null) {
    return obj;
  }

  if (Array.isArray(obj)) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-explicit-any
    return obj.map((item) => transformKeys(item, keyTransform)) as any;
  }

  const transformedObj: Record<string, T> = {};

  for (const key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      const transformedKey = keyTransform(key);
      transformedObj[transformedKey] = obj[key];
    }
  }

  return transformedObj;
};

export const camelCaseKeys = <T>(obj: Record<string, T>): Record<string, T> =>
  transformKeys(obj, (key) => key.replace(/_([a-z])/g, (_, group: string) => group.toUpperCase()));

export const snakeCaseKeys = <T>(obj: Record<string, T>): Record<string, T> =>
  transformKeys(obj, (key) => key.replace(/[A-Z]/g, (match) => `_${match.toLowerCase()}`));
