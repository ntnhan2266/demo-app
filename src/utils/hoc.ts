import React from 'react';

// Helper function to get the display name of a component
export const getDisplayName = <P>(WrappedComponent: React.ComponentType<P>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component';
};
