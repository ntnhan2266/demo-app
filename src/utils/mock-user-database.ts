import bcrypt from 'bcryptjs';
import { LOCALSTORAGE_KEY } from '@/constants/localstorage-key';
import { IUser } from '@/interfaces/user';

const getUsersFromLocalStorage = (): Map<string, IUser> => {
  const usersJson = localStorage.getItem(LOCALSTORAGE_KEY.USERS);
  const users = (usersJson ? JSON.parse(usersJson) : []) as [string, IUser][];

  // Convert the array to a map for efficient queries
  const usersMap = new Map<string, IUser>(users);
  return usersMap;
};

const saveUsersToLocalStorage = (users: Map<string, IUser>): void => {
  const usersArray = Array.from(users.entries());
  localStorage.setItem(LOCALSTORAGE_KEY.USERS, JSON.stringify(usersArray));
};

export const registerUser = (user: IUser): boolean => {
  const usersMap = getUsersFromLocalStorage();

  // Check if the email or phone is unique
  if (!usersMap.has(user.emailOrPhone)) {
    // Hash the password before saving the new user
    const hashedPassword = bcrypt.hashSync(user.password, 10);
    const newUser: IUser = {
      ...user,
      password: hashedPassword,
    };

    // Save the new user
    usersMap.set(newUser.emailOrPhone, newUser);
    saveUsersToLocalStorage(usersMap);
    return true; // Registration successful
  }

  return false; // Registration failed (email or phone is duplicated)
};

export const loginUser = (emailOrPhone: string, password: string): IUser | null => {
  const usersMap = getUsersFromLocalStorage();

  // Find the user with matching credentials
  const matchedUser = usersMap.get(emailOrPhone);

  if (matchedUser && bcrypt.compareSync(password, matchedUser.password)) {
    return matchedUser; // Return user info upon successful login
  }

  return null; // Return null for unsuccessful login
};

export const updateUser = (emailOrPhone: string, updatedInfo: Partial<Omit<IUser, 'password'>>): IUser | null => {
  const usersMap = getUsersFromLocalStorage();

  // Find the user with the specified emailOrPhone
  const userToUpdate = usersMap.get(emailOrPhone);

  if (userToUpdate) {
    // Update the user's information
    const updatedUser: IUser = {
      ...userToUpdate,
      ...updatedInfo,
    };

    // Save the updated user back to the map
    usersMap.set(emailOrPhone, updatedUser);

    // Save the updated map back to local storage
    saveUsersToLocalStorage(usersMap);

    return updatedUser; // Update successful
  }

  return null; // User not found (update failed)
};

export const updatePassword = (emailOrPhone: string, oldPassword: string, newPassword: string): boolean => {
  const usersMap = getUsersFromLocalStorage();

  // Find the user with the specified emailOrPhone
  const userToUpdate = usersMap.get(emailOrPhone);

  if (userToUpdate && bcrypt.compareSync(oldPassword, userToUpdate.password)) {
    // If the old password matches, hash the new password
    const hashedNewPassword = bcrypt.hashSync(newPassword, 10);

    // Update the user's password
    const updatedUser: IUser = {
      ...userToUpdate,
      password: hashedNewPassword,
    };

    // Save the updated user back to the map
    usersMap.set(emailOrPhone, updatedUser);

    // Save the updated map back to local storage
    saveUsersToLocalStorage(usersMap);

    return true; // Password update successful
  }

  return false; // User not found or old password does not match (password update failed)
};
