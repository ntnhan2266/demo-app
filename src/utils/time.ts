import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
dayjs.extend(utc);

/**
 * Format a timestamp or date using dayjs.
 * @param {string | number | Date} timestampOrDate - The timestamp or date to format.
 * @param {string} format - The format string (optional). Default is 'YYYY-MM-DD HH:mm:ss'.
 * @returns {string} The formatted date string.
 */
export const formatTime = (timestampOrDate: string | number | Date, format: string = 'YYYY-MM-DD HH:mm:ss'): string => {
  return dayjs(timestampOrDate).utc().format(format);
};
