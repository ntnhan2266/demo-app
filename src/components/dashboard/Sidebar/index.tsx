import React from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { useUser } from '@/hooks/useUser';
import { ROUTE_PATH } from '@/constants/route-path';
import Button from '@/components/common/Button';
import { removeAuthUserFromLocalStorage } from '@/utils/auth-storage';

const Sidebar: React.FC = (): React.ReactElement => {
  const { user } = useUser();
  const navigate = useNavigate();

  const onLogout = (): void => {
    removeAuthUserFromLocalStorage();
    navigate(ROUTE_PATH.LOGIN);
  };

  if (!user) {
    return <></>; // If there is no user, return an empty fragment
  }

  return (
    <aside className='shadow text-black p-4'>
      <div className='mb-4'>
        <h2 className='text-xl font-semibold'>
          {user.firstName} {user.lastName}
        </h2>
        <p>{user.emailOrPhone}</p>
      </div>

      <nav>
        <ul>
          <li className='mb-2'>
            <Link to={ROUTE_PATH.DASHBOARD_UPDATE_INFO} className='block w-full text-left hover:bg-green p-2'>
              Update user info
            </Link>
          </li>
          <li className='mb-2'>
            <Link to={ROUTE_PATH.DASHBOARD_UPDATE_PASSWORD} className='block w-full text-left hover:bg-green p-2'>
              Update password
            </Link>
          </li>
          <li className='mb-2'>
            <Link to={ROUTE_PATH.DASHBOARD_GITHUB_REPOS} className='block w-full text-left hover:bg-blue-300 p-2'>
              Query github repos
            </Link>
          </li>
          <li className='mb-2'>
            <Button onClick={onLogout} variant='ghost' label='Logout' fullWidth />
          </li>
        </ul>
      </nav>
    </aside>
  );
};

export default Sidebar;
