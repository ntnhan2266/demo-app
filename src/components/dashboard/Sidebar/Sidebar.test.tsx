import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Sidebar from '@/components/dashboard/Sidebar';
import { UserContext } from '@/contexts/UserContext';
import { removeAuthUserFromLocalStorage } from '@/utils/auth-storage';
import renderer from 'react-test-renderer';

jest.mock('@/utils/auth-storage', () => ({
  removeAuthUserFromLocalStorage: jest.fn(),
}));

describe('Sidebar component', () => {
  const user = {
    firstName: 'John',
    lastName: 'Doe',
    emailOrPhone: 'john.doe@example.com',
  };

  const renderComponent = () => {
    return render(
      <BrowserRouter>
        <UserContext.Provider value={{ user, setUser: jest.fn() }}>
          <Sidebar />
        </UserContext.Provider>
      </BrowserRouter>,
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
    renderComponent();
  });

  test('should match the snapshot', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <UserContext.Provider value={{ user, setUser: jest.fn() }}>
            <Sidebar />
          </UserContext.Provider>
        </BrowserRouter>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders user information when user is present', () => {
    const userNameElement = screen.getByText('John Doe');
    const userEmailElement = screen.getByText('john.doe@example.com');

    expect(userNameElement).toBeInTheDocument();
    expect(userEmailElement).toBeInTheDocument();
  });

  test('renders navigation links when user is present', () => {
    const updateInfoLink = screen.getByText('Update user info');
    const updatePasswordLink = screen.getByText('Update password');
    const githubReposLink = screen.getByText('Github repos');
    const logoutButton = screen.getByText('Logout');

    expect(updateInfoLink).toBeInTheDocument();
    expect(updatePasswordLink).toBeInTheDocument();
    expect(githubReposLink).toBeInTheDocument();
    expect(logoutButton).toBeInTheDocument();
  });

  test('calls onLogout when logout button is clicked', () => {
    const logoutButton = screen.getByText('Logout');
    fireEvent.click(logoutButton);

    expect(removeAuthUserFromLocalStorage).toBeCalled();
  });

  test('does not render anything when user is not present', () => {
    const { container } = render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: null, setUser: jest.fn() }}>
          <Sidebar />
        </UserContext.Provider>
      </BrowserRouter>,
    );

    expect(container.firstChild).toBeNull();
  });
});
