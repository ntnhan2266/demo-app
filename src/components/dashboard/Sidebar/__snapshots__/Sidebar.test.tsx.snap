// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component should match the snapshot 1`] = `
[
  <div
    className="flex justify-between items-center lg:hidden border-b border-gray-300 py-2 px-4 shadow"
  >
    <a
      className="text-black font-bold font-comfortaa text-3xl"
      href="/"
      onClick={[Function]}
    >
      βeta
    </a>
    <button
      className="cursor-pointer p-2 font-bol"
      onClick={[Function]}
    >
      <img
        alt="ic-burger-menu"
        className="w-6"
        src="ic-burger-menu.svg"
      />
    </button>
  </div>,
  <div
    className="lg:w-64 left-0 right-0 top-0 bottom-0 lg:block shadow-md transition-transform duration-300 ease-in-out p-4 bg-white z-10 lg:z-0 fixed lg:static lg:translate-x-0 translate-x-full"
    data-testid="sidebar"
  >
    <div
      className="mb-6 lg:block flex items-center justify-between"
    >
      <a
        className="text-black font-bold font-comfortaa text-3xl"
        href="/"
        onClick={[Function]}
      >
        βeta
      </a>
      <button
        className="cursor-pointer p-2 font-bol lg:hidden"
        onClick={[Function]}
      >
        <img
          alt="ic-close"
          className="w-5"
          src="ic-close.svg"
        />
      </button>
    </div>
    <h2
      className="text-xl font-semibold mb-2 text-blue-800"
    >
      John
       
      Doe
    </h2>
    <p
      className="mb-4 text-gray-600"
    >
      john.doe@example.com
    </p>
    <nav
      className=""
    >
      <ul>
        <li
          className="mb-2"
        >
          <a
            className=" block w-full text-left p-2 rounded-md transition duration-300 hover:bg-indigo-100"
            href="/dashboard/github-repos"
            onClick={[Function]}
          >
            Github repos
          </a>
        </li>
        <li
          className="mb-2"
        >
          <a
            className=" block w-full text-left p-2 rounded-md transition duration-300 hover:bg-indigo-100"
            href="/dashboard/update-info"
            onClick={[Function]}
          >
            Update user info
          </a>
        </li>
        <li
          className="mb-2"
        >
          <a
            className=" block w-full text-left p-2 rounded-md transition duration-300 hover:bg-indigo-100"
            href="/dashboard/update-password"
            onClick={[Function]}
          >
            Update password
          </a>
        </li>
      </ul>
      <button
        className="rounded-md focus:outline-none absolute left-4 right-4 bottom-6 text-white text-base px-4 py-2 bg-gray-500 hover:bg-gray-600"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Logout
      </button>
    </nav>
  </div>,
]
`;
